core-plugins/src/main/java/co/cask/hydrator/plugin/batch/file/FileCopyRecordWriter.java

```diff
@@ -121,50 +120,72 @@ public void write(NullWritable key, AbstractFileMetadata fileMetadata) throws IO
         outputStream.write(buf, 0, len);
       }
     } finally {
-      if (preserveOwner) {
-        destFileSystem.setOwner(destPath, fileMetadata.getOwner(), fileMetadata.getGroup());
+      // we have to do this to make sure even if one stream fails to close, it
+      // still attempts to close the other stream
+      try {
+        inputStream.close();
+      } finally {
+        outputStream.close();
+        // the owner is set only if the output stream is sucessfully closed
+        if (preserveOwner) {
+          destFileSystem.setOwner(destPath, fileMetadata.getOwner(), fileMetadata.getGroup());
+        }
       }
-      inputStream.close();
-      outputStream.close();
     }
   }
 
   @Override
   public void close(TaskAttemptContext taskAttemptContext) throws IOException, InterruptedException {
-    /*
-     * TODO: investigate when to close the destination filesystem and the source filesystems
-     * if we use cached connections
-     * We can't close the filesystems here because the filesystem cache is shared across
-     * multiple splits. If one split closes the connection, other splits will fail to transfer.
-     */
-    if (!fsCache) {
+    // attempts to close the other even if one fails
+    try {
       destFileSystem.close();
-      for (Map.Entry<String, FileSystem> fs : sourceFilesystemMap.entrySet()) {
-        fs.getValue().close();
+    } finally {
+      safelyCloseSourceFilesystems(sourceFilesystemMap.entrySet().iterator());
+    }
+  }
+
+  /**
+   * this method attempts to close every filesystem in the list
+   * @param fs The iterator over all the filesystems we wish to close.
+   * @throws IOException
+   */
+  private void safelyCloseSourceFilesystems(Iterator<Map.Entry<String, FileSystem>> fs) throws IOException {
+    if (fs.hasNext()) {
+      try {
+        fs.next().getValue().close();
+      } finally {
+        safelyCloseSourceFilesystems(fs);
```

Comment:
its a little weird to go through an iterator using recursion instead of a while loop. Let's just do
```
while (fs.hasNext()) {
  try {
    fs.next().close();
  } catch () {
    // log some warning
  }
}
```